packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts

```diff
@@ -0,0 +1,110 @@
+      specVersion: undefined,
+      nonce: undefined,
+      runtimeVersion: undefined,
+      signedExtensions: undefined,
+      signer: undefined,
+      tip: undefined,
+      transactionPayment: undefined,
+      transactionVersion: undefined
+    };
+  }
+
+  setNonce(value: AnyNumber){
+    this.signaturePayloadOptions.nonce = value;
+    return this;
+  }
+
+  addSignature(signer: Address | Uint8Array | string, signature: Uint8Array | string, payload: ExtrinsicPayloadValue | Uint8Array | string): GenericExtrinsic {
+    const mergeDefinedObjects = (A, B) => {
+      const res = {};
+      Object.keys({...A,...B}).map(key => {
+        res[key] = A[key] || B[key];
+      });
+      return res as ExtrinsicPayloadValue;
+    };
+
+    const mergedPayload = mergeDefinedObjects(payload, this.signaturePayloadOptions);
+    return super.addSignature(signer, signature, mergedPayload);
+  }
+
+  /**
+   * @description Sign the extrinsic with a specific keypair
+   */
+  sign(account: IKeyringPair, options: SignatureOptions): GenericExtrinsic {
+    const mergeDefinedObjects = (A, B) => {
+      const res = {};
+      Object.keys({...A,...B}).map(key => {
+          res[key] = A[key] || B[key];
+      });
+      return res as SignatureOptions;
+    };
+    const mergedSignatureOpts = mergeDefinedObjects(options, this.signaturePayloadOptions);
+    return super.sign(account, mergedSignatureOpts);
+  }
+
+  /**
+   * @description Adds payment options such as slippage and tip to the Extrinsic signature
+   */
+  paymentOpts(api: Api, feeAssetId: number | string, slippage = 0, tip = 0): this{
+    const _maxPayment = Infinity.toString();
+    const estimateFeeParams: EstimateFeeParams = {
+      encodedLength: 0, hash: undefined, isEmpty: false, registry: undefined, eq(other?: unknown): boolean {
+        return false;
+      }, toHex(isLe?: boolean): string {
+        return '';
+      }, toHuman(isExtended?: boolean): AnyJson {
+        return undefined;
+      }, toJSON(): AnyJson {
+        return undefined;
+      }, toRawType(): string {
+        return '';
+      }, toString(): string {
+        return '';
+      }, toU8a(isBare?: BareOpts): Uint8Array {
+        return undefined;
+      },
+        extrinsic : this,
+        userFeeAssetId: feeAssetId,
+        maxPayment: _maxPayment
+    }
+    const getEstimatedFee = async () => {
+      const estimatedFee = await api.derive.fees.estimateFee(estimateFeeParams)
+      return estimatedFee;
+    }
+    getEstimatedFee()
+      .then((estimatedfee) => {
+      const maxPayment = parseFloat(estimatedfee.toString()) * (1 + slippage)
+      const assetId = this.registry.createType('AssetId', feeAssetId);
+      const feeExchange = this.registry.createType('FeeExchange', { assetId, maxPayment }, 0);
+      const transactionPayment = this.registry.createType('ChargeTransactionPayment', {tip: tip, feeExchange});
+      if ('transactionPayment' in this.signaturePayloadOptions) {
```

Comment:
wondering when is this not `true` (genuine question)?

Reply:
oo just when I'm using one object to store both `SignatureOptions | ExtrinsicPayloadValue` and `ExtrinsicPayloadValue` doesn't contain transactionPayment, but thinking about it now might be cleaner just to split out the signature and payload into two objects.